### ISSUE TITLE
[explicit-member-accessibility] False positive when using private fields

### ISSUE THREAD

**Issue by doberkofler** [typescript-eslint/typescript-eslint#1666]
**Repro**

The rule `@typescript-eslint/explicit-member-accessibility` reports a false positive when using private fields. At the same time, TypeScript reports the error `An accessibility modifier cannot be used with a private identifier` if an accessibility modifier would be used.

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-member-accessibility": ["warn"]
  }
}
```

```TS
class Person {
    #name: string

    constructor(name: string) {
        this.#name = name;
    }

    greet() {
        console.log(`Hello, my name is ${this.#name}!`);
    }
}
```

**Expected Result**

No false positive when using private fields.

**Actual Result**

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.22.0` |
| `@typescript-eslint/parser`        | `2.22.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.16.1` |
| `npm`                              | `6.14.1` |

**Comment by bradzacher** [comment]
We do not yet support 3.8 syntax, including private fields.
See #1436 and associated PRs.

**Comment by jens-duttke** [comment]
@bradzacher Wouldn't it be possible to simply ignore properties starting with `#` in this rule, just like the `ignoredMethodNames` option does it for method names?

**Comment by bradzacher** [comment]
No. The AST for private names is different, and thus the `#` is not included in the name.

The AST is not yet finalised either, so we cannot build a rule on top of it.

**Comment by jens-duttke** [comment]
> the # is not included in the name.

This statement seems to be incorrect. The `util.getNameFromMember` method returns the name with the leading `#`.

I've just checked that and this small change fixed the problem:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/explicit-member-accessibility.ts
Method `checkPropertyAccessibilityModifier`.
Replace:
```
const propertyName = util.getNameFromMember(classProperty, sourceCode);
```
by
```
const propertyName = util.getNameFromMember(classProperty, sourceCode);
if (propertyName.startsWith('#')) { return; }
```

**Comment by bradzacher** [comment]
To clarify - this will work right now due to "dumb luck" of sorts.
The `getNameFromMember` function just says "if I don't know how to handle the node, just get the source code string for the node".
Hence the function returns `"#foo"`.

However - this isn't entirely correct. The name of the member is `foo`, not `#foo`.
Why does this distinction matter?

```ts
class Test {
  #foo: string = 'a';
  '#foo': string = 'b';

  test() {
    console.log(this.#foo, this['#foo']);
  }
}
```

`#foo` is a private class property whose name is `foo`.
`'#foo'` is a public class property whose name is `#foo`.

For both of these cases, however, `getNameFromMember` will currently print `#foo`.
So we cannot currently rely on this handling as it is not correct.

**Comment by colxi** [comment]
Do we have any updates on this issue?  
Unless we have a solution for this, this rule becomes useless for most projects as it will throw continuously false positives.

**Comment by LukasMachetanz** [comment]
Any updates?

**Comment by LukasMachetanz** [comment]
Is there a rough estimation when this could be fixed?

**Comment by stalniy** [comment]
> To clarify - this will work right now due to "dumb luck" of sorts.
> The `getNameFromMember` function just says "if I don't know how to handle the node, just get the source code string for the node".
> Hence the function returns `"#foo"`.
> 
> However - this isn't entirely correct. The name of the member is `foo`, not `#foo`.
> Why does this distinction matter?
> 
> ```ts
> class Test {
>   #foo: string = 'a';
>   '#foo': string = 'b';
> 
>   test() {
>     console.log(this.#foo, this['#foo']);
>   }
> }
> ```
> 
> `#foo` is a private class property whose name is `foo`.
> `'#foo'` is a public class property whose name is `#foo`.
> 
> For both of these cases, however, `getNameFromMember` will currently print `#foo`.
> So we cannot currently rely on this value to drive the lint rule as it is not correct.

I believe majority of projects don't use leading dash and quotes around class properties and even they do, now they have a problem because leading dash means a private class property in js. Even if js doesn't throw an error for this syntax, it will be confusing for developers.

So, I don't understand why all the rest should struggle to use new private class properties just for the sake of rule to be 100% complete (even though it's not 100% complete now, if we take into consideration private properties)

**Comment by HolgerJeromin** [comment]
Adding exact search term here because I stumbled over this issue, too:
> Missing accessibility modifier on class property

**Comment by OldStarchy** [comment]
I've just updated to `eslint@^8.2.0` and I'm still getting this error

|package|version|
|-|-|
|eslint|8.2.0|
|@typescript-eslint/parser|5.3.0|
|@typescript-eslint/eslint-plugin|5.3.0|
|typescript|4.3.5|
|node|16.4.0|

```
   26:2  error  Missing accessibility modifier on class property #length                     @typescript-eslint/explicit-member-accessibility
```

Do I need to do something to make this work?

**Comment by JoshuaKGoldberg** [comment]
@OldStarchy could you please post a full reproduction so that I can take a look? With just your versions and a single error there's nothing I can help with.

**Comment by jramstedt** [comment]
I'm getting `Missing accessibility modifier on class property #doesItWork.` error also.

Test case:
`
export class PrivateTest {
  #doesItWork?: number
}
`
My rule:
`
'@typescript-eslint/explicit-member-accessibility': ['error', { overrides: { constructors: 'no-public' } }]
`

**Comment by JoshuaKGoldberg** [comment]
@jramstedt, friend, fellow, internet person, please post a full reproduction including your package versions. We can't help you if you just post a couple of snippets.

**Comment by jramstedt** [comment]
Here are the versions.

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.3.1` |
| `@typescript-eslint/parser`        | `5.3.1` |
| `TypeScript`                       | `4.4.4` |
| `ESLint`                           | `8.2.0` |
| `node`                             | `16.13.0` |

Do you have all the needed info to reproduce it?

**Comment by bradzacher** [comment]
looks like it was implemented for methods
https://github.com/typescript-eslint/typescript-eslint/blob/4875cd6d224db67337f066d2a2cfb15e00587cf8/packages/eslint-plugin/src/rules/explicit-member-accessibility.ts#L113-L118

But not for properties
https://github.com/typescript-eslint/typescript-eslint/blob/4875cd6d224db67337f066d2a2cfb15e00587cf8/packages/eslint-plugin/src/rules/explicit-member-accessibility.ts#L212-L223
